File: sicp/1.11.py
def f(n):
    if n < 3:
        return n
    return f(n - 1) + f(n - 2) + f(n - 3)

def f_iter(n):
    def iter(a, b, c, counter):
        if (counter < 3):
            return c
        return iter(b, c, (a + b + c), counter - 1)
    return iter(0, 1, 2, n)

File: sicp_v2/1.11.rb
require "test/unit"

def f_recursive(n)
  if n <= 3
    return n
  end
  f_recursive(n - 1) + f_recursive(n - 2) + f_recursive(n - 3)
end

def f_iter(n)
  def iter(counter, a, b, c)
    if counter < 0
      return a
    end
    iter(counter - 1, a + b + c, a, b)
  end
  iter(n, 1, 0, -1)
end

class TestRecursiveSolution < Test::Unit::TestCase
  def test_equal
    assert_equal(3, f_recursive(3))
    assert_equal(11, f_recursive(5))
    assert_equal(20, f_recursive(6))
  end
end

class TestIterSolution < Test::Unit::TestCase
  def test_equal
    assert_equal(3, f_recursive(3))
    assert_equal(11, f_iter(5))
    assert_equal(20, f_recursive(6))
end
